### ISSUE TITLE
Exception when launching a process

### ISSUE THREAD

**Issue by myrup** [dotnet/corefx#29123]
On 2.1 preview 2 built for linux-arm (raspbian jessie), I get this after a few hours of spawning and exiting processes.
```
System.ArgumentException: An item with the same key has already been added. Key: 29932
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, Insertion
Behavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Diagnostics.ProcessWaitState.AddRef(Int32 processId, Boolean isNewChild)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
```
It's just a guess, but could it be the dictionary is not properly reaped before a new process reuses the pid ?

**Comment by danmosemsft** [comment]
@viktorhofer who was going to help stress test all this.

**Comment by tmds** [comment]
preview2 does not support having a new child Process which has the same pid as an existing Process instance (which has exited, so pid was recycled).
If you do this, then you get this exception.
Is it possible this is happening in your application?

https://github.com/dotnet/corefx/pull/27763 adds support for this. Probably that PR is in the daily builds already. Maybe you can give those a try and see what happens.

**Comment by myrup** [comment]
@tmds sounds like this is fixed by #27763 . Starting tests with the daily build to verify.

**Comment by tmds** [comment]
If there are still existing (i.e. non-Disposed) instances that refer to that pid, then we can assume we know the root cause.

On the other hand, if your application is disposing _all_ Process instances close to their exit, then there shouldn't be remaining 'state' that causes issues when a pid is recycled. This would likely mean there is a bug in the ProcessWaitState reference counting. Having https://github.com/dotnet/corefx/pull/27763 would mask that bug. As part of https://github.com/dotnet/corefx/pull/26291 I included a test for the reference counting (`TestProcessWaitStateReferenceCount`).

So it would be good to know if your app can still have Process instances for the old pid around when the new child is created.

**Comment by myrup** [comment]
Test app still runs and ready to close this issue solved by #27763 . Sidenote: Since we switched codebase from mono we're experiencing a very high cpu load after app has been up for a while. Possibly a race condition in our code, more often triggered in dotnet core. I'll investigate and open a new issue if it's not on our end.

**Comment by tmds** [comment]
> Test app still runs and ready to close this issue solved by #27763

@myrup great! time to celebrate ;) 🎉

@danmosemsft will that PR be in 2.1?

**Comment by stephentoub** [comment]
> will that PR be in 2.1?

#27763 is in the release/2.1 branch as of this morning:
https://github.com/dotnet/corefx/pull/29145/commits
